### PR TITLE
Refactoring main to move setup functions into controller/app/setup.go

### PR DESCRIPTION
**Pull request by super-harsh**
**What this PR does / why we need it**:

Our `main.go` for the controller is growing and has all the setup methods in it. In this PR, I've extracted all the code out into `controller/app/setup.go` to leave the main tidy enough.

### CODE REVIEW

File: v2/internal/crdmanagement/crds.go [hunk 1]
```diff
@@ -347,3 +353,87 @@ func VersionEqual(a apiextensions.CustomResourceDefinition, b apiextensions.Cust
 
 	return aVersion == bVersion
 }
+
+func GetNonReadyCRDs(ctx context.Context, cfg config.Values, crdManager *Manager) (map[string]apiextensions.CustomResourceDefinition, error) {
```
Comment (matthchr) [inline v2/internal/crdmanagement/crds.go:357]: Can you put these 3 functions into a separate file like `helpers` or something in this package?

Possibly can rename `crds.go` to `manager.go`, since that's really what's in here mostly?

File: v2/cmd/controller/app/setup.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+	internalflags "github.com/Azure/azure-service-operator/v2/internal/util/flags"
+	"github.com/Azure/azure-service-operator/v2/internal/util/interval"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/internal/util/lockedrand"
+	"github.com/Azure/azure-service-operator/v2/internal/version"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/conditions"
+)
+
+func SetupControllerManager(setupLog logr.Logger, flgs flags, ctx context.Context) manager.Manager {
```
Comment (matthchr) [inline v2/cmd/controller/app/setup.go:51]: `ctx` should be the first parameter here, matching Go convention?

File: v2/cmd/controller/app/setup.go [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	asometrics "github.com/Azure/azure-service-operator/v2/internal/metrics"
+	armreconciler "github.com/Azure/azure-service-operator/v2/internal/reconcilers/arm"
+	"github.com/Azure/azure-service-operator/v2/internal/reconcilers/generic"
+	internalflags "github.com/Azure/azure-service-operator/v2/internal/util/flags"
+	"github.com/Azure/azure-service-operator/v2/internal/util/interval"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/internal/util/lockedrand"
+	"github.com/Azure/azure-service-operator/v2/internal/version"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
```
Comment (matthchr) [inline v2/cmd/controller/app/setup.go:47]: I'm not totally sold on the `app/setup.go` structure here. Did you take inspiration from someplace on this `app/...` structure?

Given this stuff is all very specific to `cmd/controller` I agree it makes sense to keep here and not put into something like `pkg/setup` or `internal/setup`... do we prefer `app/setup.go` or just having these functions defined in `main.go` (or `setup.go` living alongside `main.go`)?

Basically 100% in support of new `SetupControllerManager` function, but wondering what file it makes sense to go into.

File: v2/cmd/controller/app/setup.go [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+	mgr, err := ctrl.NewManager(k8sConfig, ctrl.Options{
+		Scheme:                 scheme,
+		MetricsBindAddress:     flgs.metricsAddr,
+		NewCache:               cacheFunc,
+		LeaderElection:         flgs.enableLeaderElection,
+		LeaderElectionID:       "controllers-leader-election-azinfra-generated",
+		Port:                   9443,
+		HealthProbeBindAddress: flgs.healthAddr,
+	})
+
```
Comment (matthchr) [inline v2/cmd/controller/app/setup.go:76]: minor: remove newline here as this is an error check for the above line and so I think makes sense to not have a newline between.

File: v2/cmd/controller/app/setup.go [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+type flags struct {
+	metricsAddr          string
+	healthAddr           string
+	enableLeaderElection bool
+	crdPatterns          []string
+}
+
+func ParseFlags(args []string) (flags, error) {
```
Comment (matthchr) [inline v2/cmd/controller/app/setup.go:155]: `ParseFlags` is exported here which means that `flags` type should be too.

**Review by matthchr — COMMENTED** [review]
I like the new `Setup...` function but had some minor comments/questions around where it lives.

File: v2/cmd/controller/app/setup.go (hunk 3 above)
Comment (super-harsh) [reply v2/cmd/controller/app/setup.go:47]: I took the inspiration from how [cert-manager](https://github.com/cert-manager/cert-manager/tree/master/cmd/controller) does it and I quite liked it. We can in the future use this to add options/refactor more to make it more clean.

File: v2/cmd/controller/app/setup.go (hunk 5 above)
Comment (super-harsh) [reply v2/cmd/controller/app/setup.go:155]: Considering moving `Flags` into a separate file.

File: v2/cmd/controller/app/setup.go (hunk 3 above)
Comment (matthchr) [reply v2/cmd/controller/app/setup.go:43]: That's good enough for me.

File: v2/internal/crdmanagement/helpers.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/registration"
+)
+
+func GetNonReadyCRDs(ctx context.Context, cfg config.Values, crdManager *Manager) (map[string]apiextensions.CustomResourceDefinition, error) {
+	existingCRDs, err := crdManager.ListOperatorCRDs(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list operator CRDs")
+	}
+
+	goalCRDs, err := crdManager.LoadOperatorCRDs(CRDLocation, cfg.PodNamespace)
```
Comment (matthchr) [inline v2/internal/crdmanagement/helpers.go:29]: You might have some rebase conflicts from my recently merged "Don't list CRDs twice" PR

**Review by matthchr — APPROVED** [review]
LGTM -- might have some rebase conflicts when you merge with master though as I changed the Double-List CRD thing

File: v2/internal/crdmanagement/helpers.go (hunk 6 above)
Comment (super-harsh) [reply v2/internal/crdmanagement/helpers.go]: Yep, fixed it already. Thanks!
